plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -137,10 +140,82 @@ struct _FuGenesysUsbhubDevice {
 	FuGenesysFwCodesign codesign;
 	GByteArray *st_codesign; /* codesign info, may need to backup for GL352350 */
 	GByteArray *st_public_key;
+
+	/* hid channel */
+	FuGenesysHubhidDevice *hid_channel;
 };
 
 G_DEFINE_TYPE(FuGenesysUsbhubDevice, fu_genesys_usbhub_device, FU_TYPE_USB_DEVICE)
 
+void
+fu_genesys_usbhub_device_set_hid_channel(FuDevice *device, FuDevice *channel)
+{
+	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
+	if (self == NULL)
+		return;
+
+	if (!FU_IS_GENESYS_HUBHID_DEVICE(channel))
+		return;
+
+	if (self->hid_channel != NULL) {
+		g_warning("already set hid_channel, physical_id(%s)",
+			  fu_device_get_physical_id(FU_DEVICE(self->hid_channel)));
+	} else {
+		self->hid_channel = FU_GENESYS_HUBHID_DEVICE(channel);
+		/* align usb max length(0xffff) to usb2 packet size(0x40) */
+		self->flash_rw_size = 0xffc0;
+	}
+}
+
+static gboolean
+fu_genesys_usbhub_device_send_control_transfer(FuGenesysUsbhubDevice *self,
+					       FuGenesysUsbSetup *setup,
+					       guint8 *data,
+					       FuProgress *progress,
+					       GError **error)
+{
+	if (self->hid_channel != NULL) {
+		guint8 dir = setup->req_type.dir;
+		setup->req_type.dir = !dir; /* revert g_usb in/out dir to usb spec */
+
+		if (dir == G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST) {
+			return fu_genesys_hubhid_device_command_read(self->hid_channel,
+								     setup,
+								     data,
+								     progress,
+								     error);
+		} else {
+			if (setup->length <= GENESYS_HUBHID_REPORT_PACK_DATA_LENGTH) {
+				return fu_genesys_hubhid_device_command_write_pack(
+				    self->hid_channel,
+				    setup,
+				    data,
+				    error);
+			} else {
+				return fu_genesys_hubhid_device_command_write(self->hid_channel,
+									      setup,
+									      data,
+									      progress,
+									      error);
```

Comment:
i think this should be handled in fu_genesys_hubhid_device_command_write() when you pass in the GByteArray as a parameter.